### PR TITLE
Don't duplicate spaces from potentially-wrapped EOL-deferred lines

### LINKED ISSUE #5386: Lines that wrap on space characters at the bottom of the buffer (“new bottom line”) receive an extra line break
With 0.11.xxx selfhost builds, The following steps reliably result in visual corruption (and also drove me a little crazy!)

1. Terminal width = 120
2. Go to the terminal source tree
3. `git blame 862793299a src/types/UiaTextRangeBase.cpp`
4. Search for `GetTextRects` using `/`:
   * `/GetTextRects<CR>`
5. Go to the next match, using `n`.
6. Notice that `GetTextRects`, which should be on the top line, is _gone_.
7. Move right, left (arrow keys), and notice that it reappears.

### bad

image

### good

image

If you look at the bad image, I highlighted two things that caught my eye. Those are lines that have spaces _in column 121_, before `less` hard wraps them (!) I don't have 121 columns ...

Confirmed this in a text editor with the debug tap. I highlighted in blue the same regions that caused a break here.

The vertical strip under the cursor (look for the blue boxes) indicates column _121_. Most of the escape sequences for cursor movement start here, but in the boxed areas they're not escapes.

image

Is this the exact wrap issue? Except with conpty sending spaces instead?

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
The logic here, regarding deleting the spaces and just instantly adding
them bad, is incredibly suspect. Given that we're close to 0.11, I don't
think I can change it.

I've added a TODO with an issue number to figure out the right logic
here.

Fixes #5386.

## PR Checklist
* [x] Closes #5386
* [x] CLA signed
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I'm horrified.

## Validation Steps Performed
Tests, manual validation of the scenario in 5386 and a repro program.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
>Because we didn't remove the spaces, we shouldn't print them down here (the case under L566), because we already printed them

I would like to eventually consolidate this, because. Will we _ever_ print spaces in this branch?

```
SPACES REMOVED WHEN:
* line was not wrapped

AND

* It is optimal to prefer ECH/EL
* We are not on a new bottom line
* We did not clear all this frame
OR
* We cleared all this frame
OR
* It is a new bottom line and we are on it

SPACES RE-ADDED WHEN:
* It is not optimal to prefer ECH/EL
OR
* It is a new bottom line
OR
* We cleared all this frame

AND
* It is a new bottom line and we are on it
* The line had spaces at the end.
```

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp [hunk 1]
```diff
@@ -2359,3 +2361,57 @@ void ConptyRoundtripTests::BreakLinesOnCursorMovement()
...
+        // Buffer contents should look like the following: (80 wide)
+        // (w) means we hard wrapped the line
+        // (b) means the line is _not_ wrapped (it's broken, the default state.)
+        // cursor is on the '_'
+        //
+        // | ^ ^ ^ ^ ^ ^ ^ | ( ) (entire buffer above us; contents do not matter)
+        // |               | ( ) (entire buffer above us; contents do not matter)
+        // |AAAAAAAA...AAA | (w) (79 'A's, one space)
+        // |B_      ...    | (b) (There's only one 'B' on this line)
+
```
Comment (miniksa) [inline src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:2395]: Can you ASCII art what it looked like BEFORE too (above?)

File: src/renderer/vt/paint.cpp [hunk 2]
```diff
@@ -563,7 +563,7 @@ using namespace Microsoft::Console::Types;
         {
             _deferredCursorPos = { _lastText.X + sNumSpaces, _lastText.Y };
         }
-        else if (numSpaces > 0)
+        else if (numSpaces > 0 && removeSpaces) // if we deleted the spaces... re-add them
```
Comment (miniksa) [inline src/renderer/vt/paint.cpp:566]: Leave the "why" we're readding them.

Comment (DHowett-MSFT) [reply src/renderer/vt/paint.cpp:566]: so part of my confusion is _I actually don't know why we would re-add them_ if we chose to delete them. we should just not choose to delete them.

Comment (DHowett-MSFT) [reply src/renderer/vt/paint.cpp:566]: @zadjii-msft can you remember why we'd re-add them if we'd deleted them?

File: src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/UnitTests_TerminalCore/ConptyRoundtripTests.cpp:2395]: Sorry, I misread this test as adjusting the size after printing this in. There's no BEFORE/AFTER. Ignore.

File: src/renderer/vt/paint.cpp (hunk 2 above)
Comment (miniksa) [reply src/renderer/vt/paint.cpp:566]: Eh if we can't totally suss it out, just leave the MSFT to the code health item here.

Comment (DHowett-MSFT) [reply src/renderer/vt/paint.cpp:566]: okay, follow-up filing.

Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp:566]: I can't remember tbh. This screams organic growth to me - I think we first just decided to do ECH and removing spaces, when we didn't have wrapping (and spaces didn't matter). Then we did some deferring the cursor movement & not erasing optimizations in 20H1, which added cases where we wouldn't ECH. Now we're adding cases where we still need the spaces. I suppose if we'd written it from scratch now, accounting for wrapping, we'd do it better. 

I'd TODO it tbh.

Comment (DHowett-MSFT) [reply src/renderer/vt/paint.cpp:566]: https://github.com/microsoft/terminal/issues/5430, linking in the comments

Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp:566]: I bet we were trying to optimize out the cursor positioning that we'd do after a new line is printed. If we did optimize that scenario, and used ECH to erase some spaces, we also used CUF to move the cursor. However, if we couldn't do that, we'd still want to leave the cursor in the right place...

yea we probably should have just not removed them in the first place.

Comment (DHowett-MSFT) [inline src/renderer/vt/paint.cpp:566]:
```suggestion
        else if (numSpaces > 0 && removeSpaces) // if we deleted the spaces... re-add them
            // TODO GH#5430 - Determine why and when we would do this.
```

File: src/renderer/vt/paint.cpp [hunk 3]
```diff
@@ -563,7 +563,7 @@ using namespace Microsoft::Console::Types;
         {
             _deferredCursorPos = { _lastText.X + sNumSpaces, _lastText.Y };
         }
-        else if (numSpaces > 0)
+        else if (numSpaces > 0 && removeSpaces) // if we deleted the spaces... re-add them
         {
             std::wstring spaces = std::wstring(numSpaces, L' ');
```
Comment (DHowett-MSFT) [inline src/renderer/vt/paint.cpp:568]:
```suggestion
            // TODO GH#5430 - Determine why and when we would do this.
            std::wstring spaces = std::wstring(numSpaces, L' ');
```

**Comment by DHowett-MSFT** [comment]
@msftbot merge this in about 4 minutes

**Pull request merged**